File: git/git.go
package git

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/lguibr/comai/env"
)

func GetRepositoryName() (string, error) {
	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(out.String()), nil
}

func GetCurrentBranchName() (string, error) {
	cmd := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(out.String()), nil
}

func GetStagedChangesDiff() (string, error) {
	cmd := exec.Command("git", "diff", "--staged")
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()
	if err != nil {
		return "", err
	}
	return out.String(), nil
}

func StageAllChanges() error {
	cmd := exec.Command("git", "add", "--all")
	return cmd.Run()
}

func CreateCommit(commitMessage string) error {
	cmd := exec.Command("git", "commit", "-m", commitMessage)
	return cmd.Run()
}

func GetCommitTemplate() (string, error) {
	repoPath, err := GetRepositoryName()
	if err != nil {
		return "", err
	}

	// Construct the path to the repository-specific template file
	templatePath := filepath.Join(repoPath, "./.git/commit_template.txt")
	// Check if the file exists
	if _, err := os.Stat(templatePath); !os.IsNotExist(err) {
		templateBytes, err := os.ReadFile(templatePath)
		if err != nil {
			return "", err
		}
		template := string(templateBytes)
		return template, nil
	}

	// If no repository-specific template is found, return the global template
	return env.TEMPLATE_COMMIT, nil
}

func SaveCommitTemplate(template string) error {
	repoPath, err := GetRepositoryName()
	if err != nil {
		return err
	}

	// Construct the path to the hidden template file inside the .git directory
	templatePath := filepath.Join(repoPath, ".git", "commit_template.txt")

	// Write the template content to the file
	return os.WriteFile(templatePath, []byte(template), 0644)
}

func PerformGitReset(stepsBack int) error {
	cmd := exec.Command("git", "reset", fmt.Sprintf("HEAD~%d", stepsBack))
	return cmd.Run()
}

File: commitai/git.py
# -*- coding: utf-8 -*-
import os
import subprocess
from typing import Optional


def get_repository_name() -> str:
    return (
        subprocess.check_output(["git", "rev-parse", "--show-toplevel"])
        .strip()
        .decode()
    )


def get_current_branch_name() -> str:
    return (
        subprocess.check_output(["git", "rev-parse", "--abbrev-ref", "HEAD"])
        .strip()
        .decode()
    )


def get_staged_changes_diff() -> str:
    return subprocess.check_output(["git", "diff", "--staged"]).decode()


def stage_all_changes() -> None:
    subprocess.run(["git", "add", "--all"])


def create_commit(message: str) -> None:
    subprocess.run(["git", "commit", "-m", message])


def run_pre_commit_hook() -> bool:
    repo_path = get_repository_name()
    pre_commit_path = os.path.join(repo_path, ".git", "hooks", "pre-commit")
    if os.path.exists(pre_commit_path) and os.access(pre_commit_path, os.X_OK):
        try:
            subprocess.check_call(pre_commit_path)
            return True
        except subprocess.CalledProcessError:
            return False
    return True


def get_commit_template() -> Optional[str]:
    repo_path = get_repository_name()
    template_path = os.path.join(repo_path, ".git", "commit_template.txt")
    if os.path.exists(template_path):
        with open(template_path, "r") as f:
            return f.read()
    return os.getenv("TEMPLATE_COMMIT")


def save_commit_template(template: str) -> None:
    repo_path = get_repository_name()
    template_path = os.path.join(repo_path, ".git", "commit_template.txt")
    with open(template_path, "w") as f:
        f.write(template)


def get_unstaged_files() -> list[str]:
    """Returns a list of files that are modified/untracked but not staged."""
    # Modified files
    modified = (
        subprocess.check_output(["git", "diff", "--name-only"])
        .strip()
        .decode()
        .splitlines()
    )
    # Untracked files
    untracked = (
        subprocess.check_output(["git", "ls-files", "--others", "--exclude-standard"])
        .strip()
        .decode()
        .splitlines()
    )
    # Filter empty strings
    return [f for f in modified + untracked if f]


def stage_file(file_path: str) -> None:
    """Stages a specific file."""
    subprocess.run(["git", "add", file_path], check=True)
